src/transformers/models/groupvit/modeling_groupvit.py

```diff
@@ -0,0 +1,1864 @@
+    return attentions
+
+
+def get_grouping_from_attentions(attentions, hw_shape, rescale=False, align_corners=False):
+    """
+    Args:
+        attentions (`tuple(torch.FloatTensor)`
+    Returns:
+        grouping: list[Tensor], attention map of shape [B, groups, H, W]
+    """
+
+    attn_maps = []
+    with torch.no_grad():
+        prev_attn_masks = None
+        for idx, attn_masks in enumerate(attentions):
+            # attn_mask of shape [B, G, HxW]
+            # B: batch size (1), nH: number of heads, G: number of group token
+            if attn_masks is None:
+                assert idx == len(attentions) - 1, "only last layer can be None"
+                continue
+            # [B, G, HxW] -> [B, HxW, G]
+            attn_masks = attn_masks.permute(0, 2, 1).contiguous()
+            if prev_attn_masks is None:
+                prev_attn_masks = attn_masks
+            else:
+                prev_attn_masks = prev_attn_masks @ attn_masks
+            # [B, HxW, G] -> [B, G, HxW]
+            attn_masks = attn_masks.permute(0, 2, 1).contiguous()
+            # [B, G, HxW, G] -> [B, G, H, W]
+            cur_attn_map = resize_attn_map(prev_attn_masks, *hw_shape)
+            attn_maps.append(cur_attn_map)
+
+    # [B, G, H, W]
+    final_grouping = attn_maps[-1]
+    if rescale:
+        final_grouping = F.interpolate(final_grouping, size=hw_shape, mode="bilinear", align_corners=align_corners)
+
+    return final_grouping
+
+
+class Attention(nn.Module):
+    def __init__(
+        self,
+        config: GroupViTVisionConfig,
+    ):
+        super().__init__()
+        self.num_heads = config.num_attention_heads
+        head_dim = config.hidden_size // self.num_heads
+        self.scale = head_dim**-0.5
+
+        self.q_proj = nn.Linear(config.hidden_size, config.hidden_size)
+        self.k_proj = nn.Linear(config.hidden_size, config.hidden_size)
+        self.v_proj = nn.Linear(config.hidden_size, config.hidden_size)
+        self.proj = nn.Linear(config.hidden_size, config.hidden_size)
+
+    def forward(self, query, key=None):
+        B, N, C = query.shape
+        if key is None:
+            key = query
+        S = key.size(1)
+
+        # [B, nh, N, C//nh]
+        q = self.q_proj(query).reshape(B, N, self.num_heads, C // self.num_heads).permute(0, 2, 1, 3)
+        # [B, nh, S, C//nh]
+        k = self.k_proj(key).reshape(B, S, self.num_heads, C // self.num_heads).permute(0, 2, 1, 3)
+        # [B, nh, S, C//nh]
+        v = self.v_proj(key).reshape(B, S, self.num_heads, C // self.num_heads).permute(0, 2, 1, 3)
+
+        # [B, nh, N, S]
+        attn = (q @ k.transpose(-2, -1)) * self.scale
+        attn = attn.softmax(dim=-1)
+        assert attn.shape == (B, self.num_heads, N, S), f"{attn.shape} vs {(B, self.num_heads, N, S)}"
+
+        # [B, nh, N, C//nh] -> [B, N, C]
+        out = (attn @ v).transpose(1, 2).reshape(B, N, C)
+        out = self.proj(out)
+        return out
```

Comment:
Same comment here, please replace all one-letter variables in code/comments by more clearer things such as B->batch_size